### ISSUE TITLE
Are umap transformations non-deterministic?

### ISSUE THREAD

**Issue by greyowl** [lmcinnes/umap#158]
I am trying to use umap to preprocess some data, and I've noticed that the same vector gives a different result according to the number of rows that is being passed to the transformation. 

i.e the same row vector A outputs different vector according to the shape of the data (# of rows) being transformed. 
```python
# fit umap to data X 
reducer = umap.UMAP().fit(X)
# transform X using reducer 
embedding = reducer.transform(X)
# get subset of X to transform 
embedding_sub = reducer.transform(X[:100,:])
# => I was assuming embedding_sub == embedding[:100, :]
# => but that wasn't the case
```

**Comment by lmcinnes** [comment]
The transform is somewhat stochastic simply because of how it needs to be
implemented. Ideally the difference should be small, but I can't give any
explicit guarantees.

**Comment by assaad** [comment]
There is no way to fix the seed of the randomness used?

**Comment by lmcinnes** [comment]
There is a ``transform_seed`` parameter that can be used to fix a seed, but that can't account for subsampling the data, etc. Completely fixing the randomness is not really possible -- at least not without a lot of thought. Sorry.

**Comment by dataist** [comment]
Hey all, I thought I'd add some analysis to this discussion via colab [here](https://colab.research.google.com/drive/1Zv_iK9O5e6P8qoJRZmydJg1ECXlL0cSk). 

There I do quick quantification of the randomness introduced and explore its impact on some downstream tasks. You can find conclusions below (also in notebook):

> We see cluster assignment can be super unstable as a result of the noise introduced by `umap`'s transform even when it is the exact data used for embed training. Obviously, if your downstream task maps to what we illustrate above you may be in trouble and your downstream tasks heavily impacted by the stochastic nature of the `umap` transform...
> 
> Im curious if there are any parameters we can add to the underlying knn search to help reduce the scale of the stochastic noise it introduces (even at computational expense)?  
> 
> Separately, I would argue that the current implementation's aggressive memoization used by the transform function could obfuscate downstream problems. If its stochastic nature was evident in all scenarios it would be more obvious and help highlight when it was a source of problems like we illustrate here.
> 
> I also did some additional parameter exploration on how each may help to mitigate some of the scale of the issue (not illustrated here):
> ```
>         n_neighbors: no impact
>         n_components: no impact
>         metric: no impact
>         n_epochs: some (30%) improvement as we go over 1000
>         learning_rate: some (20%) improvement as we get around 0.5
>         init: no impact 
>         min_dist: some impact in concert with spread 
>         spread: some positive impact by setting this to np.std of X_train 
>         set_op_mix_ratio: seems best to keep at 1
>         local_connectivity: no impact,
>         transform_queue_size: no impact
>         random_state: no impact
>         transform_seed: no impact
> ```

**Comment by lmcinnes** [comment]
Thanks for the study, it will be useful for many others. In case people are interested I am working on a deterministic approach to UMAP ``transform`` functionality, but it is still very much in the "toy" stage, still has a number of drawbacks I would like to fix, and won't be ready (even if I can remedy the issues) for a while yet. Still, there is hope for now...

**Comment by dataist** [comment]
We would def be interested in contributing to a deterministic approach--we would very much like to use UMAP style embeddings in production and the current level of noise present in the transform makes that unlikely.

Also, just for context, while the colab notebook uses illustrative (but pathological) examples of how things can go wrong, I've encountered the same phenomena in real data with coherent, interpretable clusters and only being able to recover about 70% of the original cluster labels reimbedding the data used for training.

**Comment by jlmelville** [comment]
@dataist, have you looked at the effect of averaging the output of repeating the same transform multiple times?

**Comment by dataist** [comment]
@jlmelville good idea! The colab notebook analysis suggests the transformation noise is normally distributed so averaging should help eliminate it. An important caveat to that approach is that you need to not just do that for subsequent transformations but also for the original transformation returned by fitting `umap` as its just a single draw of the projection into the new space.

**Comment by apbard** [comment]
I would like to add to @dataist's experiments that we get different result for the same point also if:
- data are shuffled
- there are duplicate points (in this case it seems that just the first occurence for each duplicate assume a different embedding)

**Comment by apbard** [comment]
It would probably worse performances but to make deterministic transforms wouldn't be enough to perform a "point-by-point" transform?

**Comment by dewshr** [comment]
i am using umap in python 3.6.8, with umap-learn 0.3.1, so I am fitting the umap model with the subset of data, and when I run `umap.transform`, I am getting different transformation values for the same data points. I am using following command to create umap model
`reducer = umap.UMAP(n_components=10, random_state=0, transform_seed=0)`

**Comment by sleighsoft** [comment]
@dewshr Could you please provide a reproducible example.

**Comment by dewshr** [comment]
I have two dataset, one is subset of another dataset (subset_data and whole_data). I have also attached the sample dataset I used
```
subset_data = pd.read_csv('subset_data.csv', index_col ='name')
whole_data = pd.read_csv('whole_data.csv', index_col = 'name')

reducer = umap.UMAP(n_components=5, random_state=0, transform_seed=0)
reducer.fit(subset_data)

subset_data_transform = reducer.transform(subset_data)
whole_data_transform = reducer.transform(whole_data)

head(subset_data_transform)
name,0,1,2,3,4
var_1,-1.0818049,2.1702583,-0.7495309,-1.8568178,-1.0938863
var_2,-0.90896183,4.6964855,-0.5700368,-0.27183083,-0.59852684
var_3,-0.83147764,4.339603,0.06427337,-0.43971997,-1.0240343
var_4,-1.7820284,3.5268242,0.25737917,-1.3918203,-0.60318255
var_5,-0.0651891,2.9304326,-0.3610509,-1.4453144,-0.20545875
var_6,-1.1370527,3.421422,0.23815191,-0.5848139,-1.305079
var_7,-0.9714,2.1284223,-0.4607454,-1.8695027,-0.54702574
var_8,-1.5108252,3.919666,-0.3120325,-0.86461705,0.2995668
var_9,-1.2268223,4.120813,-1.1871228,-0.58820873,-0.15397732


head(whole_data_transform)
var_1,-1.3955795,1.7844456,-0.3540998,-1.5740595,-0.42212838
var_2,-0.6275107,5.201565,-0.7708858,-0.2549725,-0.020859836
var_3,-1.362998,4.3281274,0.8444104,-0.46012902,-0.3926726
var_4,-1.8201163,3.9629812,0.62708724,-1.1925513,0.035009027
var_5,-0.20786422,2.0100298,0.17233662,-0.9912149,-0.612357
var_6,-1.160065,3.8599873,0.5915414,0.08912127,-0.864586
var_7,-1.6482177,3.897427,-0.35188386,-1.8948933,0.69750917
var_8,-1.1465372,4.164144,-0.6294845,-1.0734795,1.0632212
var_9,-1.1563764,4.2450128,-0.48862788,0.4416442,-0.10102977

[test_data.zip](https://github.com/lmcinnes/umap/files/3611520/test_data.zip)


```

**Comment by dewshr** [comment]
> @dewshr Could you please provide a reproducible example.

did you by any chance, resolved that non-deterministic problem?

**Comment by sleighsoft** [comment]
Sorry, I have been busy lately. I will try to look at this again soon but cannot promise anything.

**Comment by qinqian** [comment]
Is there any progress in this issue? I also tried UMAP fitting with all the data, but when transform new data with the same values but adding some new points would change the embedding a little bit every time. I am using python 3.7.6 with umap 0.4.0.

**Comment by rtesselli** [comment]
I've encountered the same problem of @dewshr and others.

As @apbard suggested, I've tried by performing a "point-by-point" transform, in this way I get the correct results but the computation time explodes.

Is there any update?
